modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+  :type socket: `tcpConn`
+
+  :return: remote address
+  :rtype: `ipAddr`
+  :throws SystemError: If socket is closed.
+*/
+proc getsockname(ref socket: tcpConn): ipAddr throws {
+  var socketFd = socket.socketFd;
+  return getsockname(socketFd);
+}
+
+/*
+  See Above
+*/
+proc getsockname(ref socket: tcpListener): ipAddr throws {
+  var socketFd = socket.socketFd;
+  return getsockname(socketFd);
+}
+
+/*
+  See Above
+*/
+proc getsockname(ref socket: udpSocket): ipAddr throws {
+  var socketFd = socket.socketFd;
+  return getsockname(socketFd);
+}
+
+pragma "no doc"
+proc socket(family:IPFamily = IPFamily.IPv4, sockType:c_int = SOCK_STREAM, protocol = 0) throws {
+  var socketFd: c_int;
+  var err = sys_socket(family:c_int, sockType, protocol:c_int, socketFd);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to create socket");
+  }
+  var flags:c_int;
+  err = sys_fcntl(socketFd, F_GETFL, flags);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to get socket flags");
+  }
+  err = sys_fcntl_long(socketFd, F_SETFL, flags | O_NONBLOCK, flags);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to make socket non blocking");
+  }
+  return socketFd;
+}
+
+pragma "no doc"
+proc bind(socketFd:fd_t, ref address: ipAddr, reuseAddr = true) throws {
+  var enable = (if reuseAddr then 1 else 0):c_int;
+  setSockOpt(socketFd, SOL_SOCKET, SO_REUSEADDR, enable);
+  var err = sys_bind(socketFd, address._addressStorage);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to bind Socket");
+  }
+}
+
+
+/*
+  Bind the socket to address. The socket must not already
+  be bound to any address prior to calling this procedure.
+
+  .. code-block:: Chapel
+
+    var socket =  new udpSocket();
+    var address = ipAddr.create("127.0.0.1", 8111);
+    bind(socket, address);
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg address: address to bind to
+  :type address: `ipAddr`
+  :arg reuseAddr: whether to reuse address if already in use
+  :type address: `boolean`
+
+  :throws SystemError: If socket is closed or already bound.
+*/
+proc bind(ref socket: udpSocket, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
+
+/*
+  See Above
+*/
+proc bind(ref socket: tcpListener, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
+
+/*
+  See Above
+*/
+proc bind(ref socket: tcpConn, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
+
+pragma "no doc"
+proc naggle(socketFd:fd_t, enable:bool = true) throws {
+  var c_enable = (if enable then 0 else 1):c_int;
+  setSockOpt(socketFd, IPPROTO_TCP, TCP_NODELAY, c_enable);
+}
+
+/*
+  Enables or disables Nagle's algorithm on a given TCP Listener.
+*/
+proc naggle(ref socket: tcpListener, enable:bool = true) throws {
```

Comment:
I am expecting these to be methods on tcpListener / tcpConn. E.g. here you can write `proc ref tcpListener.nagle(enable: bool = true)`. Arguably in the design https://github.com/chapel-lang/chapel/pull/17899/files#diff-21447e19620654704318fbb1fb753236379e4d6861085d3bb4de8b7a18830199R136 it was unclear if it is a method or not.

Also in the design doc it is `nagle` rather than `naggle`

Reply:
I seemed to have used method in place of procedure but yeah I will update them to be methods

Reply:
And it should be `naggle` i will update it